source/agora/node/Registry.d

```diff
@@ -646,14 +643,25 @@ private struct ZoneData
 
          Answer a given question using this zone's data
 
+         Returns:
+          A code corresponding to the result of the query.
+          `Header.RCode.Refused` is returned for the following conditions;
+            - Query type is AXFR and query name is not matching with zone
+            - Query type is not AXFR and query name is not owned by zone
+        
+          Check `doAXFR` (AXFR queries) or `getKeyDNSRecord` (other queries)
+          for other returns.
+
     ***************************************************************************/
 
-    public void answer (in Question q, ref Message reply) @safe
+    public Header.RCode answer (in Question q, ref Message reply) @safe
     {
         if (q.qtype == QTYPE.AXFR)
-            this.doAXFR(reply);
+            return this.matches(q.qname) ? this.doAXFR(reply) : Header.RCode.Refused;
+        else if (this.owns(q.qname))
+            return this.getKeyDNSRecord(q, reply);
         else
-            reply.header.RCODE = this.getKeyDNSRecord(q, reply);
+            return Header.RCode.Refused;
```

Comment:
Bit of style guideline:
Less indentation is better (the code flows more easily). Hence, when using early returns like this, the last branch does not need to be in a `else`.